kernel/src/net/socket/netlink/route/mod.rs

```diff
@@ -0,0 +1,190 @@
+}
+
+enum Inner {
+    Unbound(UnboundNetlinkRoute),
+    Bound(BoundNetlinkRoute),
+}
+
+impl NetlinkRouteSocket {
+    pub fn new(is_nonblocking: bool) -> Self {
+        Self {
+            is_nonblocking: AtomicBool::new(is_nonblocking),
+            pollee: Pollee::new(),
+            inner: RwMutex::new(Takeable::new(Inner::Unbound(UnboundNetlinkRoute::new()))),
+        }
+    }
+
+    fn check_io_events(&self) -> IoEvents {
+        let inner = self.inner.read();
+        match inner.as_ref() {
+            Inner::Unbound(unbound) => unbound.check_io_events(),
+            Inner::Bound(bound) => bound.check_io_events(),
+        }
+    }
+
+    fn try_receive(&self, writer: &mut dyn MultiWrite, flags: SendRecvFlags) -> Result<usize> {
+        let inner = self.inner.read();
+        let bound = match inner.as_ref() {
+            Inner::Unbound(_) => {
+                return_errno_with_message!(Errno::EINVAL, "the socket is not bound")
+            }
+            Inner::Bound(bound_netlink_route) => bound_netlink_route,
+        };
+
+        bound.try_receive(writer, flags)
+    }
+}
+
+impl Socket for NetlinkRouteSocket {
+    fn bind(&self, socket_addr: SocketAddr) -> Result<()> {
+        let SocketAddr::Netlink(netlink_addr) = socket_addr else {
+            return_errno_with_message!(
+                Errno::EAFNOSUPPORT,
+                "the provided address is not netlink address"
+            );
+        };
+
+        let mut inner = self.inner.write();
+        inner.borrow_result(|owned_inner| match owned_inner.bind(&netlink_addr) {
+            Ok(bound_inner) => (bound_inner, Ok(())),
+            Err((err, err_inner)) => (err_inner, Err(err)),
+        })
+    }
+
+    fn addr(&self) -> Result<SocketAddr> {
+        let netlink_addr = match self.inner.read().as_ref() {
+            Inner::Unbound(_) => NetlinkSocketAddr::new_unspecified(),
+            Inner::Bound(bound) => bound.addr(),
+        };
+
+        Ok(SocketAddr::Netlink(netlink_addr))
+    }
+
+    fn sendmsg(
+        &self,
+        reader: &mut dyn MultiRead,
+        message_header: MessageHeader,
+        flags: SendRecvFlags,
+    ) -> Result<usize> {
+        if let Some(addr) = message_header.addr() {
+            let SocketAddr::Netlink(netlink_addr) = addr else {
+                return_errno_with_message!(
+                    Errno::EAFNOSUPPORT,
+                    "the address is in an unsupported address family"
+                );
+            };
+
+            // TODO: Further check whether other socket address can be supported.
+            if *netlink_addr != NetlinkSocketAddr::new_unspecified() {
+                return_errno_with_message!(
+                    Errno::EINVAL,
+                    "only kernel socket address is supported"
+                );
+            }
+        }
+
+        let inner = self.inner.read();
+        let bound = match inner.as_ref() {
+            Inner::Unbound(_) => todo!(),
```

Comment:
If we use something like `DgramCommon`, this can be easily implemented as `bind_to_ephemeral`, which is a correct behavior.